src/fdc3/dotnet/DesktopAgent/src/DesktopAgent/UserChannel.cs

```diff
@@ -0,0 +1,149 @@
+    {
+        public string Name { get; }
+        private readonly ConcurrentDictionary<string, string> _contexts = new ConcurrentDictionary<string, string>();
+        private string? _lastContext = null;
+        private readonly string _topicRoot;
+        private readonly IMessageRouter _messageRouter;
+        private IAsyncObserver<TopicMessage> _broadcastObserver;
+        private ILogger<UserChannel> _logger;
+
+
+        public UserChannel(string name, IMessageRouter messageRouter, ILogger<UserChannel>? logger)
+        {
+            Name = name;
+            _topicRoot = $"composeui/fdc3/userchannels/{name}/";
+            _messageRouter = messageRouter;
+
+            _logger = logger;
+        }
+
+        public async ValueTask Connect()
+        {
+            await _messageRouter.ConnectAsync();
+
+            _broadcastObserver = AsyncObserver.Create<TopicMessage>(x => HandleBroadcast(x.Payload?.ToString()));
+
+            var broadcastSubscribing = _messageRouter.SubscribeAsync(_topicRoot + "broadcast", _broadcastObserver);
+            await broadcastSubscribing;
+
+            LogConnected();
+        }
+
+        internal ValueTask HandleBroadcast(string? payload)
+        {
+            if (string.IsNullOrEmpty(payload))
+            {
+                LogNullOrEmptyBroadcast();
+                return ValueTask.CompletedTask;
+            }
+            LogPayload(payload);
+            JsonNode ctx;
+            try
+            {
+                ctx = JsonNode.Parse(payload, new JsonNodeOptions() { PropertyNameCaseInsensitive = true })!;
+            }
+            catch (JsonException)
+            {
+                LogInvalidPayloadJson();
+                return ValueTask.CompletedTask;
+            }
+            var contextType = (string?) ctx!["type"];
+
+            if (string.IsNullOrEmpty(contextType))
+            {
+                LogMissingContextType();
+                return ValueTask.CompletedTask;
+            }
+            // Although AddOrUpdate is not atomic, MessageRouter will serialize incoming broadcasts
+            // so no race condition can happen due to the non-atomic nature of AddOrUpdate
+
+            _contexts.AddOrUpdate(contextType, _ => UpdateLastContext(payload), (_, _) => UpdateLastContext(payload));
+
+            return ValueTask.CompletedTask;
+        }
+
+        private string UpdateLastContext(string context)
+        {
+            _lastContext = context;
+            return context;
+        }
+
+        internal string? GetCurrentContext(string? contextType)
+        {
+            if (contextType == null)
+            {
+                return _lastContext;
+            }
+
+            if (_contexts.TryGetValue(contextType, out var context))
+            {
+                return context;
+            }
+            return null;
+        }
+
+        private void LogConnected()
+        {
+            if (_logger?.IsEnabled(LogLevel.Debug) ?? false)
```

Comment:
Simplification: `_logger` can be non-nullable, set to `NullLogger.Instance` if not provided in the constructor.

Reply:
Done